### DISCUSSION TITLE
Issue with Task Scheduling During DST Switch

### DISCUSSION THREAD

**Discussion by HalavicH** [kagkarlsson/db-scheduler#436] [Q&A]
Hello kagkarlsson and community,

I've been using `db-scheduler` for task scheduling and have observed some behavior that I'd like to clarify.

**Current Setup**:

- Default database table configuration using provided query from example (the time is stored as local time)
- I'm using the Oracle JDBC driver for database connectivity.

**Observed Issue**: During the "Fall Back" DST transition, I noticed that tasks scheduled around the time switch can end up being scheduled for a time that's already passed.
For example, if a task finishes at 1:50 am and is set to execute every 15 minutes, it schedules itself for 1:05 am, which is in the past. This causes the task to execute immediately and continuously until the current time surpasses the DST switch.

**Additional Information**: I've looked into the `db-scheduler` code and noticed that the `OracleJdbcCustomization` class uses the default behavior from the `DefaultJdbcCustomization` class for storing `Instant` values. I'm trying to determine if the observed DST issue is due to the behavior of the Oracle JDBC driver or if it's related to how `db-scheduler` handles or configures date-time values.

**Questions**:

1. Have I possibly misconfigured something in the setup?
2. Is the observed DST behavior known, and is it related to the Oracle JDBC driver or `db-scheduler`'s handling of `Instant` values?
3. Is it possible to configure `db-scheduler` to use UTC or Unix timestamps for storing the execution time to avoid such DST-related issues?

Thank you in advance for clarification!

**Comment by HalavicH** [comment]
Well, since I use CronSchedule it's possible to pass a timezone to each cronschedule like:
new CronSchedule("* */15 * * * *", ZoneId.of("UTC"));
but is there any global setting fo it?

**Comment by kagkarlsson** [comment]
No global setting no.

This is the first time I hear about this behavior. Could it have something to do with the schema? Would be interesting to see if it could be reproduced in a test. There is an `OracleCompatibilityTest` we might experiment with..

**Reply by HalavicH** [reply]
Hi kagkarlsson,

Apologies for the delayed response, but I've delved into our issue with task scheduling during DST transitions and uncovered some noteworthy insights.

**Context:**
- We're currently using Oracle version 19.17.0.0.0.
- Our table schema closely mirrors the provided example in your documentation:

```sql
create table SCHEDULED_TASKS (
    TASK_NAME            VARCHAR2(100 char) not null,
    TASK_INSTANCE        VARCHAR2(100 char) not null,
    EXECUTION_TIME       TIMESTAMP(6),
    PICKED               NUMBER(1),
    PICKED_BY            VARCHAR2(50 char),
    LAST_SUCCESS         TIMESTAMP(6),
    LAST_FAILURE         TIMESTAMP(6),
    CONSECUTIVE_FAILURES NUMBER(19),
    LAST_HEARTBEAT       TIMESTAMP(6),
    VERSION              NUMBER(19),
    TASK_DATA            BLOB,
    primary key (TASK_NAME, TASK_INSTANCE)
)
```

**Observation:**
- In the database, task execution times are stored as strings, e.g., `2023-12-20 15:10:19.695405`, without any timezone information. Since our system operates in the Eastern European Timezone, it's susceptible to DST changes.

**Investigation:**
- I reviewed `DefaultJdbcCustomization::setInstant()` and `DefaultJdbcCustomization::getInstant()`. It appears that converting an `Instant` to a `Timestamp` omits the timezone information, resulting in the aforementioned string format.
- To address this, I experimented with adding a UTC calendar in our JDBC customization. Here's the modified code:

```java
public class DbSchedJdbcCustomization extends OracleJdbcCustomization {  
    private final Calendar utcCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"));  
  
    @Override  
    public void setInstant(PreparedStatement p, int index, Instant value) throws SQLException {  
        p.setTimestamp(index, value != null ? Timestamp.from(value) : null, utcCalendar);  
    }  
  
    @Override  
    public Instant getInstant(ResultSet rs, String columnName) throws SQLException {  
        return Optional.ofNullable(rs.getTimestamp(columnName, utcCalendar)).map(Timestamp::toInstant).orElse(null);  
    }  
}
```

**Current Status:**
- This approach seems to store timestamps in UTC (e.g., `2023-12-20 13:10:19.695405` for the example above), which is 2 hours behind EET. Preliminary testing indicates positive results, though a DST switch simulation is pending.
- I'm curious if this is a viable long-term solution, and if there are any potential oversights in this approach.
- Additionally, I'm intrigued why a timezone-aware calendar wasn't integrated into the default implementation.

**Future Consideration:**
- I'm contemplating transitioning to storing timestamps as ISO-8601 strings in the database by altering the column type to VARCHAR. Adjusting the JdbcCustomization should suffice for this change.

Would love to hear your thoughts on this approach and the current timestamp serialization method.

**Reply by kagkarlsson** [reply]
Thank you for the thorough investigation. I believe you are correct that there is something missing here. Your workaround should work. I think it might also be possible to handle by using zone-aware datatype for the column. I need to reproduce this in the compatibility tests and amend the schema (if possible)

**Reply by barbosini** [reply]
I tried HalavicHs' solution and the DST issue remains.
In my case I've scheduled a One Time Task which, upon scheduled time, get rescheduled around 55 minutes into the past

<img width="1318" height="87" alt="image" src="https://github.com/user-attachments/assets/2423e883-008a-473f-b9d1-99aeff42007e" />

<img width="1297" height="97" alt="image" src="https://github.com/user-attachments/assets/23aa500a-5654-46f3-842b-8e91f290b831" />

**Comment by barbosini** [comment]
Hi everyone

I just ran into this issue.
Was this fixed in the meantime?

Thanks
